cypress-io/cypress
# feat(npm/vue): expose Test Utils API

- Closes https://github.com/cypress-io/cypress/issues/22611
- Closes https://github.com/cypress-io/cypress/issues/22138

### User facing changelog


- Expose additional testing API from underlying Vue Test Utils library in the `cypress/vue` package.
- Inline `@vue/test-utils` type definitions, instead of relying on a peerDependency that might not be available, or the wrong version.

### Additional details


##  Expose additional testing API from underlying Vue Test Utils library in the `cypress/vue` package.
`cypress/vue` uses `@vue/test-utils` under the hood. Generally, users won't need to access this library, but there are some scenarios for advanced users where it might be useful, or where they are migrating some older jsdom + jest tests to use Cypress.

Exposing the underlying library might be useful for those scenarios. See #22611 for an examples from the community.

Generally, in Cypress, we do [global setup in a custom command](https://github.com/cypress-io/cypress/blob/1ccc90573ff0d2067db474fc6e407e446a4803e6/packages/frontend-shared/cypress/support/mock-graphql/mountFragment.ts#L23-L57). Some people may not like to use a custom command, or have some other use case in mind, so I think it's good to support alternative use cases, too.

I wouldn't consider this to be idiomatic - Cypress encourages use of custom commands where possible - so I don't think we will be recommending this in our main docs. Instead, I wrote a small note in the `@cypress/vue` README.

## Inline `@vue/test-utils` type definitions

We bundle `@vue/test-utils` entirely in the `cypress/vue` code we ship in the binary. TypeScript has no such "inline the types" feature, though, so the generated type definitions would do `import type {...} from '@vue/test-utils`.

This would give you a typing error, if you don't install a peer dependency (which we don't technically require, so it's confusing for users). In this PR, I wrote a little script to grab the types from the monorepo's `node_modules/@vue/test-utils/dist/**/*.d.ts` and paste them in `npm/vue/dist` when we build the library. This will let users get correct type definitions without needing the peer dependency. 

I also added some simple type tests to make sure it's working as expected.

### Steps to test


1. Verify problem - install cypress from npm, do `import { mount } from 'cypress/vue'` and follow it to the type def
2. You'll see they reference a library, `@vue/test-utils`, which you probably don't have installed (for example on line 3).

![image](https://user-images.githubusercontent.com/19196536/178648912-e4c82b70-3544-4602-9e45-cc4346c158cb.png)

This is a problem! Verify the fix:

1. Grab this branch
2. `cd npm/vue`
3. `yarn build`
4. Look in `cli/vue/dist/index.d.ts`

Observe it's now referring to **relative** types - note the `'./` in from of `./@vue/test-utils` on line 3:

![image](https://user-images.githubusercontent.com/19196536/178649111-0a23b7bb-7fa6-481f-9ca4-56a212cab85b.png)

As for exposing the `VueTestUtils` API, you can see the test:

https://github.com/cypress-io/cypress/pull/22757/files#diff-f9a3bc77ab97b5f10f220de6ab30201be456aad25cb85e8197c52b2b90959684

### How has the user experience changed?


### PR Tasks
